internal/core/src/indexbuilder/KmeansMajorCompaction.cpp

```diff
@@ -0,0 +1,351 @@
+            stats.add_centroids();
+        vector_field->set_dim(dim);
+        milvus::proto::schema::FloatArray* float_array =
+            vector_field->mutable_float_vector();
+        for (int j = 0; j < dim; j++) {
+            float_array->add_data(float(centroids[i * dim + j]));
+        }
+    }
+    auto output_path = file_manager_->GetCompactionResultObjectPrefix();
+
+    auto local_chunk_manager =
+        storage::LocalChunkManagerSingleton::GetInstance().GetChunkManager();
+    if (local_chunk_manager->Exist(output_path)) {
+        LOG_INFO("path {} already exists, delete it", output_path);
+        local_chunk_manager->RemoveDir(output_path);
+    }
+    local_chunk_manager->CreateDir(output_path);
+    std::string centroid_stats_path = output_path + "centroids";
+    result_files_.emplace_back(centroid_stats_path);
+    WritePBFile(stats, centroid_stats_path);
+
+    auto compute_num_in_centroid = [&](const uint32_t* centroid_id_mapping,
+                                       uint64_t start,
+                                       uint64_t end) -> std::vector<int64_t> {
+        std::vector<int64_t> num_vectors(num_clusters, 0);
+        for (uint64_t i = start; i < end; ++i) {
+            num_vectors[centroid_id_mapping[i]]++;
+        }
+        return num_vectors;
+    };
+
+    if (train_num >= data_num) {  // do not compute id_mapping again
+        uint64_t i = 0;
+        uint64_t cur_offset = 0;
+        for (auto it = insert_files.value().begin();
+             it != insert_files.value().end();
+             it++) {
+            milvus::proto::segcore::ClusteringCentroidIdMappingStats stats;
+            // write centroid_id_mapping by file sizes
+            uint64_t num_offset = offsets[i] / sizeof(T) / dim;
+            for (uint64_t j = 0; j < num_offset; j++) {
+                stats.add_centroid_id_mapping(
+                    centroid_id_mapping[cur_offset + j]);
+            }
+            cur_offset += num_offset;
+            auto num_vectors =
+                compute_num_in_centroid(centroid_id_mapping, 0, total_num);
+            for (uint64_t j = 0; j < num_clusters; j++) {
+                stats.add_num_in_centroid(num_vectors[j]);
+            }
+            std::string id_mapping_path =
+                output_path + std::to_string(it->first);
+            result_files_.emplace_back(id_mapping_path);
+            WritePBFile(stats, id_mapping_path);
+            i++;
+        }
+    } else {
+        uint64_t i = 0;
+        uint64_t start = 0;
+        uint64_t gather_size = 0;
+        // choose half of train size as a group to compute centroids
+        uint64_t group_size = train_size_new / 2;
+        std::vector<int64_t> gather_segment_id;
+        for (auto it = insert_files.value().begin();
+             it != insert_files.value().end();
+             it++) {
+            gather_segment_id.emplace_back(it->first);
+            gather_size += offsets[i];
```

Comment:
is there a special reason this has to be grouped?

how large WritePBFile could be? I'm assuming we can write everything into one file.

If stats is less than 10GB, we can parallelly read all the files and concurrently write into a same stats , and finally writePBFile to psersistent